File: TME.Store/TME.Store.Logic/ProductsProvider.cs
using System;
using System.Collections.Generic;
using System.Linq;
using TME.Api.Domain.Components;
using TME.Api.Domain.Models;
using TME.Store.Domain.Components;
using TME.Store.Domain.Models;

namespace TME.Store.Logic
{
    public class ProductsProvider : IProductsProvider
    {
        private IApiProductsProvider _apiProductsProvider;
        private IApiSearchService _apiSearchService;
        private IApiProductPricesAndStocksProvider _apiPricesAndStocksProvider;

        public ProductsProvider(IApiProductsProvider apiProductsProvider,
                                IApiProductPricesAndStocksProvider apiPricesAndStocksProvider,
                                IApiSearchService apiSearchService)
        {
            _apiProductsProvider = apiProductsProvider;
            _apiPricesAndStocksProvider = apiPricesAndStocksProvider;
            _apiSearchService = apiSearchService;
        }

        public SearchProductsResult Search(string searchText, int page)
        {
            ApiSearchResult apiSearchResult = _apiSearchService.Search(searchText, page);
            List<string> listOfSymbols = apiSearchResult.ProductList.Select(s => s.Symbol).ToList();

            ApiPriceResult<ApiProductPriceAndStock> productPrices = _apiPricesAndStocksProvider.GetPricesAndStocks(listOfSymbols);

            List<Product> products = (from apiProduct in apiSearchResult.ProductList
                                      join apiProductPriceAndStock in productPrices.ProductList on apiProduct.Symbol equals apiProductPriceAndStock.Symbol
                                      select new Product(
                                                apiProduct.Symbol,
                                                apiProduct.Producer,
                                                apiProduct.Description,
                                                apiProduct.Category,
                                                apiProduct.Unit,
                                                apiProduct.Photo,
                                                apiProduct.Thumbnail,
                                                apiProductPriceAndStock.Amount,
                                                 new ProductPrice(
                                                    apiProductPriceAndStock.VatRate,
                                                    apiProductPriceAndStock.VatType,
                                                    apiProductPriceAndStock.PriceList.Select(p =>
                                                                        new PriceOffer(
                                                                        p.Amount,
                                                                        p.PriceValue,
                                                                        p.Special)).ToList())))
                                                .ToList();

            int pageCount = (int)Math.Ceiling(apiSearchResult.Amount / 20.0);

            return new SearchProductsResult(productPrices.PriceType,
                                            productPrices.Currency,
                                            products,
                                            pageCount);
        }
    }
}

File: tme_store/lib/components/products_provider.dart
import 'package:tme_store/api/components/products_repository.dart';
import 'package:tme_store/api/models/models.dart';
import 'package:queries/collections.dart';
import 'package:tme_store/models/models.dart';

class ProductsProvider {
  ProductsRepository _repository;

  ProductsProvider(this._repository);

  Future<SearchProductsResult> search(String text, int page) async {
    if(text.isEmpty) {
      return SearchProductsResult.empty();
    }

    var searchResult = await _repository.search(text, page);
    
    if(searchResult.totalProductsFound == 0) {
      return SearchProductsResult.empty();
    }

    var priceResult = await _repository
        .getPricesAndStocks(searchResult.products.map((p) => p.symbol).toList());

    var searchProductsCollection = Collection(searchResult.products);
    var pricesCollection = Collection(priceResult.products);

    var products = pricesCollection.join<ApiProduct, String, Product>(
        searchProductsCollection,
        (p1) => p1.symbol,
        (p2) => p2.symbol,
        (p1, p2) => Product(
            amount: p1.amount,
            symbol: p2.symbol,
            category: p2.category,
            description: p2.description,
            producer: p2.producer,
            weight: p2.weight,
            unit: p2.unit,
            photoUrl: p2.photoUrl,
            thumbnailUrl: p2.thumbnailUrl,
            price: ProductPrice(
                vatRate: p1.vatRate,
                vatType: p1.vatType,
                priceOffers: p1.prices.map((apiPrice) => PriceOffer(
                    ammountInOffer: apiPrice.amount,
                    isSpecialPrice: apiPrice.isSpecial,
                    price: apiPrice.value)).toList()))).toList();

    return SearchProductsResult(
      currency: priceResult.currency, 
      priceType: priceResult.priceType,
      products: products,
      totalCount: searchResult.totalProductsFound);
  }
}
